### DISCUSSION TITLE
How to add apps to the x86_64 image?

### DISCUSSION THREAD

**Discussion by MikeDev101** [joshiemoore/snakeware#165] [Q&A]
I've been messing around with snakeware for a while on the Raspberry Pi 4. With a bit of help from a spare Linux machine, adding code is relatively easy. As for the x86_64 build, not so much.

I've tried compiling from source but it keeps failing (either it complains about the makefile having an invalid configuration or some binaries still using 32-bit when building for 64-bit), so I gave up on that. I did some digging and (much to my dismay) all of the snakeware code is stored in an initrd image, which is a pain to unpack, not to mention repackage.

I also tried messing with the snakeuser partition, which yielded similar results.

Is there a way to add new code? It would help a lot if someone pointed me in the right direction (or at least, gave me a clue on how to do so).

**Comment by joshiemoore** [comment]
Hey, so I just pushed a commit which updates buildroot to the latest version and solves the build issues people were having. If you pull that then you should be able to build for x86-64 without problems.

If you add your app to the appropriate directory in `snakewm/apps/` in this repo then it will get automatically packed into the image by the post-build script along with the other apps.

Feel free to post if you have any other problems, I'm looking forward to seeing what you're working on!

**Comment by MikeDev101** [comment]
> Hey, so I just pushed a commit which updates buildroot to the latest version and solves the build issues people were having. If you pull that then you should be able to build for x86-64 without problems.

Ok, thanks.

> If you add your app to the appropriate directory in `snakewm/apps/` in this repo then it will get automatically packed into the image by the post-build script along with the other apps.

I was hoping for a way to add apps directly onto a drive with snakeware imaged onto it (as if I were to drag-drop code into snakeuser, and use a bit of "from x import y" from there?), so I can quickly debug code.

**Comment by joshiemoore** [comment]
The snakeware ISO partition is intended to be "ROM" which persists unmodified between boots. But what you could do is set up a persistent partition as shown in this video: https://www.youtube.com/watch?v=qBbr9N1TIqQ

Then copy the directory containing your app to the persistent partition. Once you boot into snakeware, you could then copy your app's directory from the persistent partition to `/usr/lib/python3.9/snakewm/apps/<category>`. Then when you launch snakewm, the app should appear in the app menu.

This won't persist between boots of course, only changes made in the snakeuser partition will. Maybe we could add a feature which automatically loads apps from a "userapps" folder in the persistent partition on boot to make it easier for people to load apps that aren't packaged in the snakeware ROM, that would be cool.

Edit to note: I haven't tried copying apps from the snakeuser partition to the `snakewm/apps` directory, so I'm not 100% certain this will work, but I have a feeling it should just work. If you try it could you post here with the results?

**Reply by MikeDev101** [reply]
> Once you boot into snakeware, you could then copy your app's directory from the persistent partition to `/usr/lib/python3.9/snakewm/apps/<category>`. Then when you launch snakewm, the app should appear in the app menu.

Something like 
> import os
> os.system("cp /snakeuser/<source> /usr/lib/python39/snakewm/apps/<category>")

I presume.
This should work, but it's not the most elegant system. Oh well.
Thanks for the tip.

> The snakeware ISO partition is intended to be "ROM" which persists unmodified between boots. But what you could do is set up a persistent partition as shown in this video: https://www.youtube.com/watch?v=qBbr9N1TIqQ

I'll look into that.

> This won't persist between boots of course, only changes made in the snakeuser partition will. Maybe we could add a feature which automatically loads apps from a "userapps" folder in the persistent partition on boot to make it easier for people to load apps that aren't packaged in the snakeware ROM, that would be cool.

Perhaps that could be a feature suggestion? I don't see it too difficult to write a python script that runs on startup that copies whatever it finds in something like "/snakeuser/import" to that directory in a ramdisk?

> Edit to note: I haven't tried copying apps from the snakeuser partition to the `snakewm/apps` directory, so I'm not 100% certain this will work, but I have a feeling it should just work. If you try it could you post here with the results?

I'll either do an edit to this reply or make a new reply.

Edit: I forgot to mention: my machine threw up all over the place while compiling. It might be my Linux distro. I tried compiling it on an x86-64 machine running Raspberry Pi OS Desktop, which runs on top of Debian Buster...
